좋아요 기능 추가

hossi97 opened the pull request:
- 포스트에 대해 좋아요 등록/삭제 기능 추가함

## 📝 세부사항
- 좋아요 기능 `refactoring` 및 변경된 접근제어 수정함
- PreAuthorize를 사용하여 접근제어 구현함
- CurrentUser를 사용하여 사용자에 관한 정보 가져옴
- 기존의 status변경하는 방식이 아닌 DB에서 데이터를 삭제하는 방식으로 변경함
- Optional<Like> like = likeRepository.findByUserAndPost(user, post);  => orElseGet()으로 로직 수정함

## 📗 참고사항
[JPA의 delete](https://velog.io/@park2348190/JPA%EC%9D%98-delete-%EC%BF%BC%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C#%EA%B2%B0%EB%A1%A0)
[JPA의 save](https://wangtak.tistory.com/2)
[JPQL vs nativeQuery](https://velog.io/@youmakemesmile/Spring-Data-JPA-JPQL-%EC%82%AC%EC%9A%A9-%EB%B0%A9%EB%B2%95Query-nativeQuery-DTO-Mapping-function)
[@Query vs nativeQuery = true](https://wonit.tistory.com/470)
[좋아요 구현 참고](https://velog.io/@00300x00501/Chapter-8-%EC%A2%8B%EC%95%84%EC%9A%94-%EA%B5%AC%ED%98%84%ED%95%98%EA%B8%B0#%EC%A2%8B%EC%95%84%EC%9A%94-%EA%B5%AC%ED%98%84---%EC%A2%8B%EC%95%84%EC%9A%94-%EA%B5%AC%ED%98%84-%EC%99%84%EB%A3%8C)

## 📎 관련 이슈

hossi97 committed:
feat: 포스트 목록 조회 기능 추가

- 포스트 목록 조회 기능 추가
- 좋아요한 포스트 목록 조회 기능 추가

hossi97 committed:
feat: 좋아요 등록/삭제 기능 추가

- 포스트에 대해 좋아요 등록/삭제 기능 추가

Nitoriubar commented on src/main/java/com/triprint/backend/domain/like/service/LikeService.java:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	@Transactional
+	public void activateLike(Like like) {
+		like.changeStatus(true);
+		likeRepository.save(like);
+	}
+
+	@Transactional
+	public void deactivateLike(Like like) {
+		like.changeStatus(false);
+		likeRepository.save(like);
+	}
```
register과 unresiter과 별개로 만드신 이유가 있을까요?

Nitoriubar committed:
Merge branch 'dev' into feat/like

-Conflicts 수정완료

Nitoriubar committed:
refactor: 변경된 접근제어 수정 및 리팩토링

- PreAuthorize 수정
- CurrentUser 수정
- active/inactive로 매서드 이름 변경
- Like객체를 찾거나 새로 만들 경우의 로직 수정
- Like객체를 찾아올 경우, UserId와 PostId로 찾아오는 방식으로 수정
  (repository에서 에러 발생, 해결중..)

Nitoriubar committed:
refactor: UserId와 PostId로 좋아요 생성하거나 찾아오는 방식으로 변경완료

- JPQL이 아닌 SQL을 사용하여 변경완료함(NativeQuery = true)

Nitoriubar committed:
refactor : 좋아요 On / Off 구현 완료

- 로그인한 사용자가 좋아요한 게시물들만 가져오는 기능 구현
- 전체 게시물 불러오는 기능 구현
- 좋아요 On/Off 구현

Nitoriubar committed:
refactor : 좋아요 Off할 경우 likeRepository에서 삭제하는 방식으로 변경

sangrimlee commented on src/main/java/com/triprint/backend/domain/like/controller/LikeController.java:
```diff
@@ -0,0 +1,38 @@
+	@PutMapping("/{postId}")
+	@PreAuthorize("hasRole('ROLE_USER')")
+	public ResponseEntity<LikeDto> deactivateLike(@CurrentUser UserPrincipal userPrincipal, @PathVariable Long postId) {
+		likeService.unregisterLike(userPrincipal.getId(), postId);
+
+		return ResponseEntity.ok(likeService.unregisterLike(userPrincipal.getId(), postId));
```
likeService.unregisterLike가 두 번 사용되요!

Nitoriubar replied:
넵, 수정했습니다!

Nitoriubar committed:
fix : 중복된 코드 삭제

sangrimlee reviewed (commented):
고생하셨습니다👍

Nitoriubar committed:
fix : LikeDto의 이름을 LikeResponse로 변경하여 정확하게 구분해줌
